src/bitcoin_fmt_plug.c

```diff
@@ -331,19 +323,15 @@ static int crypt_all(int *pcount, struct db_salt *salt)
 #if ARCH_LITTLE_ENDIAN==1
 			for (i = 0; i < sizeof(key)/sizeof(uint64_t); i++)  // the derived key
 				((uint64_t *)key)[i] = JOHNSWAP64(key_iv[SIMD_COEF_64*i + (index2&(SIMD_COEF_64-1)) + index2/SIMD_COEF_64*SHA_BUF_SIZ*SIMD_COEF_64]);
-			for (i = 0; i < sizeof(iv)/sizeof(uint64_t); i++)   // the derived iv
-				((uint64_t *)iv)[i]  = JOHNSWAP64(key_iv[SIMD_COEF_64*(sizeof(key)/sizeof(uint64_t) + i) + (index2&(SIMD_COEF_64-1)) + index2/SIMD_COEF_64*SHA_BUF_SIZ*SIMD_COEF_64]);
 #else
 			for (i = 0; i < sizeof(key)/sizeof(uint64_t); i++)  // the derived key
 				((uint64_t *)key)[i] = key_iv[SIMD_COEF_64*i + (index2&(SIMD_COEF_64-1)) + index2/SIMD_COEF_64*SHA_BUF_SIZ*SIMD_COEF_64];
-			for (i = 0; i < sizeof(iv)/sizeof(uint64_t); i++)   // the derived iv
-				((uint64_t *)iv)[i]  = key_iv[SIMD_COEF_64*(sizeof(key)/sizeof(uint64_t) + i) + (index2&(SIMD_COEF_64-1)) + index2/SIMD_COEF_64*SHA_BUF_SIZ*SIMD_COEF_64];
 #endif
 
 			AES_set_decrypt_key(key, 256, &aes_key);
-			AES_cbc_encrypt(cur_salt->cry_master, output, cur_salt->cry_master_length, &aes_key, iv, AES_DECRYPT);
+			AES_cbc_encrypt(cur_salt->cry_master + cur_salt->cry_master_length - 32, output, 32, &aes_key, iv, AES_DECRYPT);
 
-			if (check_pkcs_pad(output, cur_salt->cry_master_length, 16) == 32) {
```

Comment:
This was a bug, right? AFAIK `check_pkcs_pad(data, len, blocksize)` can never return > `blocksize` so this was always false prior to this commit.

Reply:
Yep. We have this bug in two places.

Apparently, this wrong check was working in practice! Maybe Bitcoin wallets have two padding blocks? Otherwise, I can't seem to explain how this check used to succeed.

Reply:
Ah, no this was not a bug. The return (when not -1) is not the *padding* but the data length w/o padding. 32 was correct before, 16 is correct now.

Reply:
That explains things, thanks!